app/services/concerns/foreman_puppet/extensions/bulk_hosts_manager.rb

```diff
@@ -0,0 +1,26 @@
+module ForemanPuppet
+  module Extensions
+    module BulkHostsManager
+      extend ActiveSupport::Concern
+
+      def change_puppet_proxy(proxy_id, ca_proxy)
+        proxy = SmartProxy.find_by(id: proxy_id)
+        begin
+          @hosts.each do |host|
+            if ca_proxy
+              host.puppet_ca_proxy = proxy
+            else
+              host.puppet_proxy = proxy
+            end
+            host.save(:validate => false)
```

Comment:
Do we need callbacks to be run or could we take a shortcut and assign the proxy with a single update query?